toolboxes/core/cpu/hoNDArray_utils.h

```diff
@@ -276,123 +276,271 @@ namespace Gadgetron {
+          if (val == T(0)){
+              memset(out_ptr, 0, out->get_number_of_bytes());
+          }
+          else{
+                size_t N = out->get_number_of_elements();
+                long long n;
+                #pragma omp parallel for default(none) private(n) shared(N, out_ptr, val)
+                for (n = 0; n<(long long)N; n++)
+                {
+                    out_ptr[n] = val;
+                }
+          }
+      }
 
-    for( size_t frame_offset=0; frame_offset<num_batches; frame_offset++ ){
-#ifdef USE_OMP
-#pragma omp parallel for
-#endif
-      for( long long idx=0; idx<(long long)num_elements_out; idx++ ){
-        const typename uint64d<D>::Type co_out = idx_to_co<D>( idx, matrix_size_out );
-        T _out;
-        bool inside = (co_out>=offset) && (co_out<(matrix_size_in+offset));
+      typename uint64d<D>::Type matrix_size_in = from_std_vector<size_t, D>(*in->get_dimensions());
+      typename uint64d<D>::Type matrix_size_out = from_std_vector<size_t, D>(*out->get_dimensions());
 
-        if( inside )
-          _out = in_ptr[co_to_idx<D>( co_out-offset, matrix_size_in)+frame_offset*num_elements_in];
-        else{
-          _out = val;
-        }
-        out_ptr[idx+frame_offset*num_elements_out] = _out;
+      if (weak_greater(matrix_size_in, matrix_size_out)){
+          throw std::runtime_error("pad: size mismatch, cannot expand");
       }
-    }
+
+      typename uint64d<D>::Type offset(D);
+      for (d = 0; d<D; d++){
+          offset[d] = matrix_size_out[d]/2 - matrix_size_in[d]/2;
+      }
+
+      size_t len = in->get_size(0);
+      size_t num = in->get_number_of_elements() / len;
+
+      long long k;
+
+#pragma omp parallel default(none) private(k, d) shared(in_ptr, out_ptr, num, len, in, out, offset)
+      {
+          std::vector<size_t> ind;
+
+#pragma omp for 
+          for (k = 0; k < (long long)num; k++){
+              ind = in->calculate_index(k*len);
+              for (d = 0; d < D; d++){
+                  ind[d] += offset[d];
+              }
+
+              T* out_ptr_curr = out_ptr + out->calculate_offset(ind);
+              memcpy(out_ptr_curr, in_ptr + k*len, sizeof(T)*len);
+          }
+      }
+  }
+
+  template<class T> void
+  pad(size_t x, hoNDArray<T> *in, hoNDArray<T>* out, bool preset_out_with_val = true, T val = T(0))
+  {
+      typename uint64d<1>::Type padSize(1);
+      padSize[0] = x;
+      pad<T, 1>(padSize, in, out, preset_out_with_val, val);
+  }
+
+  template<class T> void
+  pad(size_t x, size_t y, hoNDArray<T> *in, hoNDArray<T>* out, bool preset_out_with_val = true, T val = T(0))
+  {
+      typename uint64d<2>::Type padSize(2);
+      padSize[0] = x;
+      padSize[1] = y;
+      pad<T, 2>(padSize, in, out, preset_out_with_val, val);
+  }
+
+  template<class T> void
+  pad(size_t x, size_t y, size_t z, hoNDArray<T> *in, hoNDArray<T>* out, bool preset_out_with_val = true, T val = T(0))
+  {
+      typename uint64d<3>::Type padSize(3);
```

Comment:
This does not do what you think it does. Should be
vector_td<size_t,3> padSize(x,y,z);
What you're doing is simply filling a vector of size 3 with the number 3.